### ISSUE TITLE
Copy-paste doesn't work on GNU/Linux, from another program to Minetest

### ISSUE THREAD

**Issue by Calinou** [minetest/minetest#373]
It only seems to work on Windows. Copy-paste from Minetest to Minetest works (example: select something in chat, press Ctrl+C, then press Ctrl+V).

**2014-08-15:** tested, the bug is still here.

**Comment by Novatux** [comment]
Bug still present, here is what I got by testing:
- If you copy text inside Minetest, you will be able to paste it both inside Minetest and to other programs; however, after pasting in another program, you won't be able to paste it in Minetest again
- Copying text from another program will prevent you from pasting text inside Minetest; including text that was previously copied from inside Minetest.
That means that doing after any copy-paste action (copying or pasting) from a program outside Minetest, you won't be able to paste text into Minetest until you select text and copy it from Minetest.

Finally, trying to copy text from one Minetest instance to another crashes the first one (the source), while doing nothing on the second.
This is the error that the first one gets:
X Error of failed request:  BadAtom (invalid Atom parameter)
  Major opcode of failed request:  18 (X_ChangeProperty)
  Atom id in failed request:  0x0
  Serial number of failed request:  798
  Current serial number in output stream:  800
AL lib: (EE) alc_cleanup: 1 device not closed

**Comment by PilzAdam** [comment]
@Novatux I have a totally different behavior here:
* Copying text *from* Minetest doesn't work at all; it only fills my clipboard with an empty string.
* Copying text *to* Minetest from another program does only work after I press Ctrl+V once (all following presses paste the text correctly).
* I cannot reproduce the crash that you get

I use KDE SC 4.5.8

**Comment by Novatux** [comment]
FYI, I use i3wm, with debian jessie

**Comment by yo-we** [comment]
Got this quite similar, but not exactly the same (on Linux Mint LMDE + minetest 0.4.12 "mainline"):
 No paste from outside Minetest has any effect whatsoever inside the game (tested in client tab, T-chat and F10-chat).
 Text copied from inside Minetest game (I can only copy stuff before starting the game, like server address or user name) can be pasted correctly (tested in client tab, T-chat and F10-chat).
 Quite annoying, as the copy paste was supposed to be a work around for Minetest chat not accepting some basic characters like à, Umlaut, etc.

By the way: It would be very usefiul to be able to paste directly in the T-chat, once this is fixed (not only in the F10-chat).

**Comment by est31** [comment]
Still reproducible with latest irrlicht from svn, [bug report here](https://sourceforge.net/p/irrlicht/bugs/435).

**Comment by est31** [comment]
Submitted a fix to upstream : http://irrlicht.sourceforge.net/forum/viewtopic.php?f=7&t=51075

**Comment by ShadowNinja** [comment]
@est31: I had trouble applying that (tabs converted to four spaces, carriage returns stripped).  [Here's one that applies to 1.8.3](http://ix.io/mo7/diff) (`wget http://ix.io/mo7 -O- | git apply`)

**Comment by est31** [comment]
Ok, it might be that I've done the change against an earlier version, where it was space indented, or thanks to git converting tabs to spaces, idk.
Either way, its only a three-liner, I'll note it though on the forum.

About a workaround: unfortunately it isn't possible, as one would have to access a private member, see this try of me: https://github.com/est31/minetest/commit/1d07b2e580f1ed80625c6c91a65f9d822d22b202

**Comment by 0-afflatus** [comment]
Understood that this is an irrlicht problem.
On GNU/Linux:
1) Minetest should accept Ctrl+C/Ctrl+V copy/paste et al; and middle-click pasting of the primary selection. 
2) Text fields should also accept AltGr+ combinations for typing characters outside the user's keyboard layout. (Is that called keyboard compositing?)
Is 2 a separate issue AFA Irrlicht and this BR are concerned?

**Comment by ShadowNinja** [comment]
This seems to be fixed in Irrlicht 1.8.3, can someone else confirm?

**Comment by nerzhul** [comment]
It's fixed in irrlicht 1.9 i can confirm that. Closing as it's an upstream issue

**Comment by est31** [comment]
Its still existent for me in 1.8.1 and I haven't heard that it was fixed in 1.8.3, it requires a st

**Comment by paramat** [comment]
Closing as fixed upstream.

**Comment by est31** [comment]
@paramat only if you use the latest unstable branches of irrlicht, there is no release yet with the fix.

**Comment by ShadowNinja** [comment]
@est31: So we just keep this issue open waiting for Irrlicht to make the release?  What's the point of keeping it open?

**Comment by est31** [comment]
If its not reaching any users except those who got the development version of irrlicht from svn (probably there is even an arch repo for this), I don't think we should close this. I mean I could have closed this issue when I wrote the patch, because I can say "I know a patch that fixes it", without sharing it with the world, it has the same effect on the end user.

Idk, its some kind of an arbitrary choice when to say "okay now we can assume everybody can get it working in a reasonable amount of time", only with updating software.

We could be really mean, and say that on ubuntu precise, a platform we support, an irrlicht version is used that doesn't have this bug fixed, and therefore we can only close the bug if precise (and all other distros) get EOLed. So saying "it can be closed when irrlicht releases a working stable version" is some kind of compromise here.

Perhaps we could simply take the "upstream issue" label as the reason for the close, so automatically close all upstream labeled issues, that's a good concept imo, but it would affect more issues than just this.

**Comment by 0-afflatus** [comment]
If there is no more work or discussion needed this end, you may as well close. This is really about your workflow. It can always be re-opened when the time comes.

**Comment by ShadowNinja** [comment]
I think this should be closed as soon as a patch is merged into the upstream project because:
  * It's useful to have until that time, to remind people that it's an issue that needs fixing, should someone want to write a patch.  It's still useful while a patch is proposed but not merged, because the patch still needs review and might be rejected.
  * After then it just becomes extra maintenance work: You have to wait until a stable release and then remember to close the issue.  During this time it isn't helpful either: if someone is looking for something to fix they may see this, but they'll waste their time looking at it because it's already fixed and there's nothing they can do (other than perhaps push the upstream developers to release a new version sooner).

**Comment by est31** [comment]
And the functionality may be reverted before the stable release comes out because it violates someone's (workflow)[https://xkcd.com/1172/], or something.

Probably many bugs we have with irrlicht are fixed in STK's fork. Should we close all issues with irrlicht because probably they are fixed in the STK fork? Even if its only a theory?

**Comment by est31** [comment]
Okay, 1.8.4 is released, let's close it: http://irrlicht.sourceforge.net/forum/viewtopic.php?f=2&t=51448

Anybody still experiencing this bug: please update your irrlicht.

**Comment by ExeterDad** [comment]
@est31 Nice work fixing upstream :+1:

**Comment by ClaudiusMinimus** [comment]
As a user of the ppa for the daily build on Mint 18 64-bit, updating Irrlicht is not an issue, because I don't compile this build. On a daily basis I run the Mint Update Manager to update MT. For me this is a security issue, because I use a password locker app for all my username/passwords and would like to use its copy/paste functions for entering my password each time I log in to a server. I have tried the double ctrl+v workaround with no success. The lack of the ability to paste from my password locker is a security issue also because I am either forced to painstakingly enter secure passwords by hand or use less secure password which are easy to enter. I hope you will be able to fix this login security bug. 
Thank you,
ClaudiusMinimus

**Comment by HybridDog** [comment]
l think the clipboard is handled by the X server.
What about the stk's fix?

**Comment by est31** [comment]
STK forked irrlicht, they can do with it whatever they want. And their fork is GPL licensed now, we can't use it.

**Comment by Megaf** [comment]
Bug resolved with Irrlicht 1.8.4 indeed.

**Comment by Megaf** [comment]
Bug still present, just morphed into something else.
I can copy things from other applications and paste with Ctrl C in fomspecs. But I can't paste them on the  game console (F10), If we still had our old chat formspec paste would work there but it is not working on our current chat thing, since it's based on the terminal thing.

Should we reopen this or should I open a new issue?

**Comment by HybridDog** [comment]
in terminal you press shift+ctrl+v to paste

**Comment by ClaudiusMinimus** [comment]
When you say, "in terminal" if you mean the system terminal, then yes, I can paste text, but unless I can run the MT client and connect to a server (including my username and password) from a command line, I do not see how this is useful. The issue is from the MT gui login page that I am unable to paste using any key combinations suggested, such as:
1. shift+ctrl+v
2. ctrl+v
3. ctrl+v twice in a row

Thank you,
ClaudiusMinimus

**Comment by est31** [comment]
Mhhh, seems the PPA statically links irrlicht.

@ClaudiusMinimus if you look at `/etc/apt/sources.list` there should be a line with the minetest PPA url. Change the "[Distribution part](https://wiki.debian.org/SourcesList)" to read "yakkety", and then run apt-get update and apt-get dist-upgrade. Maybe that fixes the bug, but if you see an error, revert back to prevent it from affecting the rest of your system (an invalid sources.list file breaks the apt mechanism). If you can't find the line in `/etc/apt/sources.list` itself, try to look at files in `/etc/apt/sources.list.d/`.

Either way, there is no bug anymore.

**Comment by HybridDog** [comment]
yakkety doesn't work for me:
![2016-10-15-110850_1920x1080_scrot](https://cloud.githubusercontent.com/assets/3192173/19408892/f5377ea8-92c7-11e6-96af-a6117cce2dc7.png)

**Comment by ClaudiusMinimus** [comment]
@est31 , Thank you for your detailed instructions. I had no problems understanding or implementing any of them. However, copy/paste into the password login field still does not work.

>  Mhhh,  seems the PPA statically links irrlicht.

I made the following changes:

1. Updated /etc/apt/sources.list.d/:
`# deb http://ppa.launchpad.net/minetestdevs/daily-builds/ubuntu xenial main
# deb-src http://ppa.launchpad.net/minetestdevs/daily-builds/ubuntu xenial main
deb http://ppa.launchpad.net/minetestdevs/daily-builds/ubuntu yakkety main
deb-src http://ppa.launchpad.net/minetestdevs/daily-builds/ubuntu yakkety main` 

2. Successfully ran: apt-get update and apt-get dist-upgrade
3. Attempted to copy/paste in password field (failed)
4. Rebooted: 2nd attempted to copy/paste in password field (failed)
5. Updated to 4.4.0-43 kernel for LM 18 (xenial)
6. Rebooted: 3rd attempted to copy/paste in password field (failed)

In all attempts I used each of the combinations of paste that has been previously discussed.

> Either way, there is no bug anymore.

It seems to me that since LinuxMint 18 LTS is built on Ubuntu 16.04 LTS that this "bug" is still active for a major distribution of Linux. I have not tested on Ubuntu, but my guess is that any distro built on xenial will fail pasting into the Minetest client login fields.

Thank you for your help,
ClaudiusMinimus

**Comment by est31** [comment]
@ClaudiusMinimus can you run minetest from the commandline and pass the --info param to it? The output should show the irrlicht version. If it shows something older than irrlicht 1.8.4 it is not recent.

> this "bug" is still active for a major distribution of Linux.

That's due to how linux distributions with stable releases work, and not our fault in any way. We fixed the bug, and if the distros don't adopt the fix soon enough for you, you have to talk with them and not with us, or consider different options, like:

1. using a rolling release distro.
1. manually compile irrlicht + minetest
1. get [the flatpak package](https://forum.minetest.net/viewtopic.php?f=42&t=15336). You'll probably have to install flatpak for that to work.

Either way, this is more of a user support discussion, and not about this bug, which is fixed now. Probably its better discussed on the forums.

**Comment by ClaudiusMinimus** [comment]
@est31  thank you for the kind suggestions. I greatly appreciate the respectful way you have addressed this issue. 
> can you run minetest from the commandline and pass the --info param to it? 

I can see from the --info flag that my Irrlicht Engine version 1.8.3 is too old. I am grateful to have identified the cause of this issue and see that the issue continues due to my distro's limitations.

```
2016-10-16 09:20:41: INFO[Main]: Irrlicht: Irrlicht Engine version 1.8.3
2016-10-16 09:20:41: INFO[Main]: Irrlicht: Linux 4.4.0-43-generic #63-Ubuntu SMP Wed Oct 12 13:48:03 UTC 2016 x86_64
2016-10-16 09:20:41: INFO[Main]: Irrlicht: Using renderer: OpenGL 3.0
2016-10-16 09:20:41: INFO[Main]: Irrlicht: Mesa DRI Intel(R) Sandybridge Mobile : Intel Open Source Technology Center
2016-10-16 09:20:41: INFO[Main]: Irrlicht: OpenGL driver version is 1.2 or better.
2016-10-16 09:20:41: INFO[Main]: Irrlicht: GLSL version: 1.3

```
I will try your other suggestions and let you know if they worked or not, whidh is for the benefit of other users who may discover this thread.

Also, if I continue this discussion I will start a new thread on the forums as requested and post the link to it here.

Thank you,
ClaudiusMinimus

**Comment by ClaudiusMinimus** [comment]
> get the [flatpak package](https://forum.minetest.net/viewtopic.php?f=42&t=15336). You'll probably have to install flatpak for that to work.

Your instructions were clear and exact - It works perfectly!  

Thank you!
ClaudiusMinimus

**Comment by rdococ** [comment]
From my tests on Debian, I can't copy from other applications into Minetest, but anything I copy from Minetest I can paste anywhere - outside of Minetest or in it with no restrictions. Copying from one Minetest session to another clears the clipboard and forces the copied session to quit.

**Comment by red-001** [comment]
I'm using Irrlicht 1.8.4 and I'm still experiencing this bug.

**Comment by growlf** [comment]
Hmm...  this apears to still (several years later?) be an issue on Ubuntu 16.x through Ubuntu 19.10 vanilla installaitions with Minetest as the first and only directly installed package (beyond basic requirements).  I cannot copy past into/outof the game screen.  I can, of course, paste into the server console and chat from there, but that is tedious and not particularly effective for real-world needs.

**Comment by rubenwardy** [comment]
This issue is superseded by this one: https://github.com/minetest/minetest/issues/7830

I will say that it's unlikely that Irrlicht will ever fix this issue, and it may be worth working around it/fixing it ourselves somehow
